File: test-fixtures/languages/c/complex/deep_flow_sink.c
/* deep_flow_sink.c -- Final executor with dangerous sink.
 *
 * Receives resolved command strings and executes them via system().
 * The taint chain terminates here at system() and popen().
 *
 * All types use "Dfc" prefix. Standard library only.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "deep_flow_chain.h"

/* Step 29: Initialize executor config */
void dfc_executor_init(struct DfcExecConfig *config) {
    strncpy(config->shell, "/bin/bash", sizeof(config->shell) - 1);
    config->shell[sizeof(config->shell) - 1] = '\0';
    config->timeout_sec = 60;
}

/* Step 30: Build environment prefix */
static void dfc_build_env_prefix(const char *environment, char *prefix,
                                 size_t prefix_sz) {
    if (strcmp(environment, "production") == 0) {
        prefix[0] = '\0';
    } else {
        snprintf(prefix, prefix_sz, "echo [%s] && ", environment);
    }
}

/* Step 31: Combine prefix with command */
static void dfc_combine_command(const char *prefix, const char *command,
                                char *full, size_t full_sz) {
    snprintf(full, full_sz, "%s%s", prefix, command);
}

/* Step 32: Wrap in shell invocation */
static void dfc_wrap_in_shell(const char *shell, const char *command,
                              char *wrapped, size_t wrapped_sz) {
    snprintf(wrapped, wrapped_sz, "%s -c '%s'", shell, command);
}

/* Step 33: Trim whitespace from command */
static void dfc_trim_command(char *command) {
    /* Trim leading whitespace */
    char *start = command;
    while (*start == ' ' || *start == '\t') {
        start++;
    }
    if (start != command) {
        memmove(command, start, strlen(start) + 1);
    }
    /* Trim trailing whitespace */
    size_t len = strlen(command);
    while (len > 0 && (command[len - 1] == ' ' || command[len - 1] == '\t')) {
        command[--len] = '\0';
    }
}

/* Step 34: Execute with popen and capture output */
static struct DfcExecResult dfc_execute_with_capture(const char *command) {
    struct DfcExecResult result;
    memset(&result, 0, sizeof(result));
    strncpy(result.command, command, sizeof(result.command) - 1);

    /* SINK: tainted data reaches popen() */
    FILE *proc = popen(command, "r");
    if (proc == NULL) {
        result.exit_code = -1;
        strncpy(result.output, "popen failed", sizeof(result.output) - 1);
        return result;
    }

    char line[256];
    result.output[0] = '\0';
    while (fgets(line, sizeof(line), proc) != NULL) {
        strncat(result.output, line,
                sizeof(result.output) - strlen(result.output) - 1);
    }

    result.exit_code = pclose(proc);
    return result;
}

/* Step 35: Execute with system() as fallback */
static struct DfcExecResult dfc_execute_direct(const char *command) {
    struct DfcExecResult result;
    memset(&result, 0, sizeof(result));
    strncpy(result.command, command, sizeof(result.command) - 1);

    /* SINK: tainted data reaches system() */
    result.exit_code = system(command);
    strncpy(result.output, "(direct execution)", sizeof(result.output) - 1);
    return result;
}

/* Steps 29-35: Main executor entry point */
struct DfcExecResult dfc_executor_run(const char *command,
                                      const char *environment,
                                      struct DfcExecConfig *config) {
    /* Step 30: Build environment prefix */
    char prefix[128];
    dfc_build_env_prefix(environment, prefix, sizeof(prefix));

    /* Step 31: Combine prefix with command */
    char full_command[1024];
    dfc_combine_command(prefix, command, full_command, sizeof(full_command));

    /* Step 32: Wrap in shell invocation */
    char wrapped[1024];
    dfc_wrap_in_shell(config->shell, full_command, wrapped, sizeof(wrapped));

    /* Step 33: Trim */
    dfc_trim_command(wrapped);

    printf("[EXEC] Running: %s\n", wrapped);

    /* Step 34: Try captured execution first */
    struct DfcExecResult result = dfc_execute_with_capture(wrapped);
    if (result.exit_code != 0) {
        /* Step 35: Fall back to direct execution */
        fprintf(stderr, "[WARN] Captured execution failed, trying direct\n");
        result = dfc_execute_direct(wrapped);
    }

    return result;
}

File: test-fixtures/languages/php/complex/deep_flow_sink.php
<?php
/* deep_flow_sink.php -- Final stage: format and execute the command.
 * Tainted user input from $_GET reaches exec()/shell_exec() after
 * flowing through the full pipeline. All types prefixed with Dfc.
 */

/* Steps 26-29: DfcCommandFormatter -- wraps command for shell execution */
class DfcCommandFormatter {
    /* Step 26: Add timeout wrapper */
    private function addTimeout(DfcExecutionContext $ctx): void {
        $ctx->formatted_command = 'timeout ' . $ctx->timeout_sec . ' ' .
                                  $ctx->formatted_command;
    }

    /* Step 27: Add directory change prefix */
    private function addDirectoryPrefix(DfcExecutionContext $ctx): void {
        $ctx->formatted_command = 'cd ' . $ctx->working_dir . ' && ' .
                                  $ctx->formatted_command;
    }

    /* Step 28: Escape for shell (incomplete -- does not handle all cases) */
    private function shellEscape(string $cmd): string {
        $escaped = str_replace('"', '\\"', $cmd);
        return $escaped;
    }

    /* Step 29: Wrap in shell invocation */
    private function addShellWrapper(DfcExecutionContext $ctx): void {
        $escaped = $this->shellEscape($ctx->formatted_command);
        $ctx->formatted_command = '/bin/sh -c "' . $escaped . '"';
    }

    public function format(DfcExecutionContext $ctx): bool {
        if (empty($ctx->formatted_command)) {
            return false;
        }
        $this->addTimeout($ctx);
        $this->addDirectoryPrefix($ctx);
        $this->addShellWrapper($ctx);
        return true;
    }
}

/* Steps 30-32: DfcPipelineExecutor -- runs the command */
class DfcPipelineExecutor {
    /* Step 30: Execute with exec() and capture output */
    private function executeCaptured(string $command): int {
        $output = [];
        $return_code = 0;
        /* SINK: tainted data reaches exec() */
        exec($command, $output, $return_code);
        foreach ($output as $line) {
            echo "[OUTPUT] " . $line . "\n";
        }
        return $return_code;
    }

    /* Step 31: Execute with shell_exec() as fallback */
    private function executeDirect(string $command): string|false {
        /* SINK: tainted data reaches shell_exec() */
        $result = shell_exec($command);
        return $result;
    }

    /* Step 32: Try captured, fall back to direct */
    public function execute(DfcExecutionContext $ctx): int {
        if ($ctx->dry_run) {
            echo "[DRY RUN] Would execute: " . $ctx->formatted_command . "\n";
            return 0;
        }

        echo "[EXEC] Running: " . $ctx->formatted_command . "\n";

        $result = $this->executeCaptured($ctx->formatted_command);
        if ($result !== 0) {
            error_log("[WARN] Captured execution failed, trying direct");
            $direct_output = $this->executeDirect($ctx->formatted_command);
            if ($direct_output !== false) {
                echo "[OUTPUT] " . $direct_output;
                return 0;
            }
            return 1;
        }

        return $result;
    }
}
